types/wlr_foreign_toplevel_management_v1.c

```diff
@@ -0,0 +1,583 @@
+		bool state) {
+	struct wlr_foreign_toplevel_handle_v1 *toplevel =
+		toplevel_handle_from_resource(resource);
+	if (!toplevel) {
+		return;
+	}
+	struct wlr_foreign_toplevel_handle_maximized_event event = {
+		.toplevel = toplevel,
+		.maximized = state,
+	};
+	wlr_signal_emit_safe(&toplevel->events.request_maximize, &event);
+}
+
+void foreign_toplevel_handle_set_maximized(struct wl_client *client,
+		struct wl_resource *resource) {
+	toplevel_handle_send_maximized_event(resource, true);
+}
+
+void foreign_toplevel_handle_unset_maximized(struct wl_client *client,
+		struct wl_resource *resource) {
+	toplevel_handle_send_maximized_event(resource, false);
+}
+
+static void toplevel_send_minimized_event(struct wl_resource *resource,
+		bool state) {
+	struct wlr_foreign_toplevel_handle_v1 *toplevel =
+		toplevel_handle_from_resource(resource);
+	if (!toplevel) {
+		return;
+	}
+
+	struct wlr_foreign_toplevel_handle_minimized_event event = {
+		.toplevel = toplevel,
+		.minimized = state,
+	};
+	wlr_signal_emit_safe(&toplevel->events.request_minimize, &event);
+}
+
+static void foreign_toplevel_handle_set_minimized(struct wl_client *client,
+		struct wl_resource *resource) {
+	toplevel_send_minimized_event(resource, true);
+}
+
+static void foreign_toplevel_handle_unset_minimized(struct wl_client *client,
+		struct wl_resource *resource) {
+	toplevel_send_minimized_event(resource, false);
+}
+
+static void foreign_toplevel_handle_activate(struct wl_client *client,
+		struct wl_resource *resource, struct wl_resource *seat) {
+	struct wlr_foreign_toplevel_handle_v1 *toplevel =
+		toplevel_handle_from_resource(resource);
+	if (!toplevel) {
+		return;
+	}
+
+	struct wlr_seat_client *seat_client = wlr_seat_client_from_resource(seat);
+	struct wlr_foreign_toplevel_handle_activated_event event = {
+		.toplevel = toplevel,
+		.seat = seat_client->seat,
+	};
+	wlr_signal_emit_safe(&toplevel->events.request_activate, &event);
+}
+
+static void foreign_toplevel_handle_close(struct wl_client *client,
+		struct wl_resource *resource) {
+	struct wlr_foreign_toplevel_handle_v1 *toplevel =
+		toplevel_handle_from_resource(resource);
+	if (!toplevel) {
+		return;
+	}
+	wlr_signal_emit_safe(&toplevel->events.request_close, toplevel);
+}
+
+static void foreign_toplevel_handle_set_rectangle(struct wl_client *client,
+		struct wl_resource *resource, struct wl_resource *surface,
+		int32_t x, int32_t y, int32_t width, int32_t height) {
+	struct wlr_foreign_toplevel_handle_v1 *toplevel =
+		toplevel_handle_from_resource(resource);
+	if (!toplevel) {
+		return;
+	}
+
+	if (width < 0 || height < 0) {
+		wl_resource_post_error(resource,
+			ZWLR_FOREIGN_TOPLEVEL_HANDLE_V1_ERROR_INVALID_RECTANGLE,
+			"invalid rectangle passed to set_rectangle: width/height < 0");
+		return;
+	}
+
+	struct wlr_foreign_toplevel_handle_set_rectangle_event event = {
+		.toplevel = toplevel,
+		.surface = wl_resource_get_user_data(surface),
```

Comment:
It's better to never use plain `wl_resource_get_user_data` like this. There are type-safe helper functions available, here `wlr_surface_from_resource`.

Reply:
Yes, thanks for the suggestion. I had even used `wl_surface` instead of `wlr_surface` ...